packages/@lwc/engine/src/framework/api.ts

```diff
@@ -347,10 +359,24 @@ export function c(
     Ctor: ComponentConstructor,
     data: CustomElementCompilerData,
     children?: VNodes
-): VCustomElement {
+): VCustomElement | VElement {
     if (isCircularModuleDependency(Ctor)) {
         Ctor = resolveCircularModuleDependency(Ctor);
     }
+    if (isFunction(Ctor) && Ctor.prototype instanceof HTMLElement) {
+        // the resolved reference for this component is a Custom Element class
+        if (process.env.NODE_ENV !== 'production') {
+            data.validateFn = function validateCustomElement(this: VElement) {
+                const { sel, elm } = this;
+                if (!(elm instanceof Ctor)) {
+                    throw new Error(
+                        `Custom element ${Ctor} must be registered as "${sel}" via customElements.define().`
+                    );
+                }
+            };
+        }
+        return h(sel, data, children || EmptyArray);
```

Comment:
this is the meat of it, if the constructor is a real custom element, it must be registered (but we don't have a way to check that here), so, instead we add a `validateFn` to vnode, to check that after `document.createElement()` is invoked by the diffing algo. In this case, we don't want to return a VCustomElement, we want just a VElement and let the browser to deal with everything.